### ISSUE TITLE
Suggestion for radar icons

### ISSUE THREAD

**Issue by artfoundry** [scottschiller/ArmorAlley#13]
I'd like to suggest an option to increase the size of the radar blips/icons - on my phone they're very hard to distinguish or survey quickly (which you often need to do while flying; half the time i can't even see the dot for my helicopter when I'm moving).

**Comment by scottschiller** [comment]
Thanks for the note, I'll take a look. In portrait orientation, there isn't much space to render the 8192-pixel-wide battlefield at scale. The entire battlefield needs to be shown, and that means units need to be pretty small in order to fit without being a mass of overlapping blobs.

I'm sure there's a better approach, but I don't have anything off the top of my head at the moment. This will need some more thinking.

**Comment by scottschiller** [comment]
Good thinking - and yes, the turrets and the missile launchers are the ones I like to keep an eye out for! 😅 

I was thinking of scaling, and how to fit everything. e.g., I made the radar 1.25x or 1.5x bigger, would it help? If the radar took up more vertical space, I could make everything taller - but it doesn't necessarily help the restriction on width.

If I simply magnify the current UI, things are more "legible" but with some trade-off as currently displayed. You can't see the entire world, now the enemy base is off-screen - so imagine everything crunched a little tighter together, probably more overlap. I think this is worth exploring, it feels like there's a good balance somewhere in here. With larger scale, maybe I can look at using borders instead of filling in the colour - or, show small versions of the actual sprites. TBD!

Here's a quick mock on desktop - note that the enemy base isn't shown as I applied a "zoom" to everything.
<img width="1409" alt="AA - 1 5x radar mock-up" src="https://github.com/scottschiller/ArmorAlley/assets/174437/48fc906d-1228-482f-b42d-d9ed124c115b">

This approach also takes away from the battlefield height a bit, so the game will be scaled down a little in size. I don't think this will be an issue on desktop, as it'd be another 32 to 64 pixels off the window height; it would be more notable on small portrait screens, e.g., an iPhone.

With a larger radar, I could also remove the bottom bit of the UI with the "order queue" and options / exit links on small screens to maximize vertical space. Maybe I find a new home for those.

**Comment by artfoundry** [comment]
I think not seeing the other base is an ok tradeoff, as I'm usually not paying much attention to that anyway unless I'm flying on that side of the battlefield.  You could even allow the user to drag the radar to view the rest (and then it resets as soon as they start moving again).

As for taking up more vertical space, I'd actually be ok with moving the exit and options buttons - i actually often get the options button and the unit ordering/helicopter controls switch/button confused, so moving it would be great!  You could even put the exit button inside the Options panel.

Btw, i think you have a spelling error in one of the messages - i think it said "You stated capturing a bunker".

**Comment by scottschiller** [comment]
This is now an experimental pref, enabled and scaling active by default for mobile. On desktop, you can use `s` to toggle scaling in-game. On mobile, tap-and-hold with a second finger while flying will do the same.

1.5x feels like it gives a bit more room, without giving up too much. The smallest portrait screens make this one a challenge. 😅 
https://github.com/scottschiller/ArmorAlley/assets/174437/751a1dc4-fbfb-465d-adb5-387ab5ae3ccc

**Comment by scottschiller** [comment]
I think this is an improvement. Spacing things out further, but maintaining icon dimensions. 1x vs. 2x vs. 3x, portrait and landscape. Portrait looks to benefit most from horizontal scaling, and maybe that's where I apply this to start.

I think to make things work similarly on landscape, I also need to make the icons larger in addition to scaling and scrolling.

iPhone portrait, 1x / 2x / 3x:
![iPhone portrait, 1x](https://github.com/scottschiller/ArmorAlley/assets/174437/beb5ca0b-6c40-45f8-820f-872b717caf34)
![iPhone portrait, 2x](https://github.com/scottschiller/ArmorAlley/assets/174437/8b8b35bb-ecd9-40d1-8d24-ff90e3eb05c2)
![iPhone portrait, 3x](https://github.com/scottschiller/ArmorAlley/assets/174437/31f8778d-47ff-41f8-87b7-118e62683fc4)

iPad portrait, 1x / 2x / 3x:
![iPad portrait, 1x](https://github.com/scottschiller/ArmorAlley/assets/174437/00aacf9e-02b4-48d1-95a9-f49e5614943e)
![iPad portrait, 2x](https://github.com/scottschiller/ArmorAlley/assets/174437/053dd696-394b-4217-aaaa-4352ab601c89)
![iPad portrait, 3x](https://github.com/scottschiller/ArmorAlley/assets/174437/0f3940a7-6b22-48b3-a517-824c1836b04d)

iPhone landscape, 1x / 3x:
<img width="894" alt="iPhone landscape, 1x" src="https://github.com/scottschiller/ArmorAlley/assets/174437/82b606aa-3f87-4714-82b2-0cbd9ee56c01">
![iPhone landscape, 3x](https://github.com/scottschiller/ArmorAlley/assets/174437/70757825-7b90-421e-8648-6a040f99cc5d)

iPad landscape, 1x / 3x:
<img width="1286" alt="iPad landscape, 1x" src="https://github.com/scottschiller/ArmorAlley/assets/174437/080e2d4d-705c-445b-a28f-a23cd00ad2f5">
<img width="1286" alt="iPad landscape, 3x" src="https://github.com/scottschiller/ArmorAlley/assets/174437/98c4aa72-d729-433b-8180-8df5947a3936">

**Comment by artfoundry** [comment]
I haven't tried on ipad, but looks good on my iphone! My only request would be to either brighten the icon for they player chopper or not have it blink - it's still kinda hard to see at a glance on my phone in landscape (does anyone actually play it in portrait??), even with the zoomed radar.

**Comment by scottschiller** [comment]
Thanks for the continued feedback; I've made a big round of updates and tweaks to the UI over the last few weeks.
I've introduced a radar scrollbar / rangefinder(?) of sorts that sits on the fuel bar UI, so you can roughly gauge your position and range / visibility on the battlefield. This UI gets slightly different treatment for contrast / visibility on smaller landscape screens e.g., iPhone, I may make this an option in prefs.

I've also made the scale and icons larger, helicopter now blinks more "clearly" instead of fading - so it can still get your attention - and radar icons are a bit taller on small landscape screens as well.

I think it's looking pretty good, on the various devices and simulators I've got on hand.
<img width="888" alt="iPhone, landscape" src="https://github.com/scottschiller/ArmorAlley/assets/174437/8a270a79-9031-4c68-b9d1-647ee11b3ce0">
<img width="447" alt="iPhone, portrait" src="https://github.com/scottschiller/ArmorAlley/assets/174437/26a029e9-35c6-45a3-adfe-d8e761bb51fe">
<img width="1286" alt="iPad, landscape" src="https://github.com/scottschiller/ArmorAlley/assets/174437/230823e6-2034-488f-9a95-032d70cb84e2">

**Comment by artfoundry** [comment]
Looking pretty good!  But it looks like the range indicators in the radar are stuck "on", even if the checkboxes are unchecked.

<img width="1003" alt="Screenshot 2024-01-02 at 11 19 19 PM" src="https://github.com/scottschiller/ArmorAlley/assets/4854419/675b8382-9107-4ad5-a9e3-af85f585adea">

**Comment by scottschiller** [comment]
@artfoundry Thanks for the bug report - I wasn't applying the CSS change to the radar element in the DOM, hah. Now fixed.

I've basically refactored the radar to use `<canvas>` for rendering, as well as the battlefield effects (the starfield, gunfire, shrapnel, and "confetti.") The result should be a significant drop in CPU, reduction of jank and improved smooth framerate / gameplay. I've got more in the works.

There are some new radar details like chains on balloons, blinking lights on landing pads, angle / rotation on smart missiles and so forth.

I've tweaked the radar scaling and layout to be more responsive, as well. Radar elements can be more or less fixed-height, and scale in terms of being more "spaced out" across the battlefield for the most part vs. being stretched.

A few screenshots follow.

Desktop chrome, full-screen (4K):
<img width="2560" alt="Desktop Chrome, full-screen @ 4K" src="https://github.com/scottschiller/ArmorAlley/assets/174437/78f48fec-2f31-45a1-8eb8-5d13c110500d">

iPad landscape, radar @ 1X + 3X:
<img width="1286" alt="iPad, landscape - radar @ 1x" src="https://github.com/scottschiller/ArmorAlley/assets/174437/949c4878-6bee-4072-8ea7-3c2bca80a2bc">
<img width="1286" alt="iPad, landscape - radar @ 3x" src="https://github.com/scottschiller/ArmorAlley/assets/174437/b95aa31f-e4bb-4680-8660-9d95494e7fdc">

iPhone, radar @ 3X (portrait) and 2X (landscape)
<img width="447" alt="iPhone, portrait - radar @ 3x" src="https://github.com/scottschiller/ArmorAlley/assets/174437/995b7f77-5f5f-4fdd-b9ec-29534cacb882">
<img width="888" alt="iPhone, landscape - radar @ 2x" src="https://github.com/scottschiller/ArmorAlley/assets/174437/348949a5-3d98-4e82-8bd8-31155eb7fa82">

**Comment by artfoundry** [comment]
Nice.  I also found a small spelling error in the last battle completion letter (finally finished the single player campaign on easy! 😅).
<img width="342" alt="Screenshot 2024-01-03 at 10 38 26 PM" src="https://github.com/scottschiller/ArmorAlley/assets/4854419/c071145a-fef1-4ec4-900d-5f1a7a5a97dc">

**Comment by scottschiller** [comment]
Ha! Good catch, thanks. I remember hurriedly typing all those out at once, looking at the handbook and flipping through the pages to find each of the letters. It took some time with all the HTML formatting, too.

**Comment by artfoundry** [comment]
Oh those were part of the original game instructions?  I can't remember whether I had the manual, much less what was in it if I did. I just figured you made those up.

**Comment by scottschiller** [comment]
Ha! So the "old tanker" backstory was from the printed manual, not my work. I originally got a copy of the game from a friend (as I suspect many did,) in maybe 1991 or 1992, and I learned how to play the game before ever seeing the manual - referred to as the "combat handbook," which turned out to be quite well done.

You can find a pretty nice scan of the manual here. https://archive.org/details/armoralleymanual/ArmorAlley-Manual/
It definitely helped clarify a few things for me in 2013, when I got my first official boxed copy of the game from eBay. It doesn't explain everything, and encourages exploration. I think there are a good number of undocumented features, like the benefit of owning all bunkers in earlier battles etc.

The manual says the game doesn't have any copy protection, but you are expected to register with a serial number that's printed on the label of the disk(s). If you play a network game using copies registered with the same serial number, you're limited to a "demonstration" game (maybe a demo battle / level?), and can't order extra helicopters.

I now have three boxed copies of the game from eBay; two PC versions, and one for the classic Macintosh. One of the PC versions is on two 5.25" floppies, pretty classic stuff. The others are a single 3.5" floppy. I suspect the 5.25" version was an early release, as the combat handbook also used "wiro" or "plastic coil" binding - a nice little detail - whereas the others are stapled.

It's been fun digging through the original boxed software and combat handbook, pulling details and other bits for the remake. There are still a few things I could add and work on ("dumb" missiles and napalm, save / resume battle, badges, the scoring system and high scores), but I feel like I have the vast majority of the original covered - and very few people have noticed any omissions!

**Comment by artfoundry** [comment]
That's awesome. I'll take a look at the manual later tonight.  I thought I remembered there being napalm in later missions!
I think I mentioned this either in another issue or in a response on a YouTube video - I did beta testing for the original developer like 15 years ago when he was trying to port the game to iOS. Unfortunately he gave up even though he was so close.  Glad you didn't!

**Comment by artfoundry** [comment]
Ha, I just saw in the manual that Three-Sixty used to be right around the corner from where I live now!

**Comment by scottschiller** [comment]
I'm a bit jealous you got to try the iOS beta! I discovered some [mentions of the iOS port](https://www.youtube.com/watch?v=HLWBPkNe48k) (YouTube) when starting my remake in 2013. Mostly, I wanted to see if there was still commercial viability at the time, or was the game more "abandonware" and so forth. I found the author and reached out, and was very thankful to receive permission to use the name and graphics for my version. I don't recall the exact reply, but at some point I sent a link along and got a "Hey, that's pretty good!" or similar. 🙇‍♂️ 

It's funny, and a small world. I remembered Bascom Avenue from when I played the game as a kid, and I eventually moved to Sunnyvale and then SF for work in tech. I would go record digging in the South Bay (Streetlight in San Jose, "Big Al's Record Barn", etc.) somewhat regularly and was quite delighted when I saw the Bascom Avenue / Campbell exit off... 280, I think it was. ;)

**Comment by artfoundry** [comment]
That's awesome he approved, both legally and appreciatively.

Ha, my favorite spot for buying music was Tower Records in Mountain view. I'm still amazed that Rasputin Music is still open on Bascom in Campbell, but I guess vinyl has made a comeback. Are you still in the bay area?

**Comment by scottschiller** [comment]
Ah yes; I forgot about Rasputin, that was the other place I'd stop at.
I had a great time in California, but returned to Canada close to five years ago.

**Comment by scottschiller** [comment]
Work in progress: Making the radar icons more detailed and identifiable.
The tank shape was a fun one.
![radar-detail-3](https://github.com/scottschiller/ArmorAlley/assets/174437/6f483f2e-9ee5-40b6-99bb-83dd99bd6778)
![radar-detail-2](https://github.com/scottschiller/ArmorAlley/assets/174437/292fcc3b-bdd3-4bd5-98f9-a9bf5e9c5a92)
![radar-detail-1](https://github.com/scottschiller/ArmorAlley/assets/174437/f1f311a6-d088-44f0-ba61-5fe0f5be639c)

**Comment by scottschiller** [comment]
Radar design tweaks, ongoing: Dropping the diagonal stripes in favour of a simple gradient. Bunkers and super bunkers now have doorways. Scan nodes use a drop shadow, same as on the battlefield. Turrets were tightened up a bit, they were almost overlapping when in groups of three on some battles.

I think this is much improved over previous designs, and I can probably do more. Thanks for starting this discussion last November!
![Radar gradient, scan range inset shadows, bunker details](https://github.com/scottschiller/ArmorAlley/assets/174437/8fd8cc17-83fb-42e1-bba8-d5aaebaab3fd)
![Radar - balanced game](https://github.com/scottschiller/ArmorAlley/assets/174437/db1b68bc-5e70-4293-8b73-b4ce6ccb3d23)

**Comment by artfoundry** [comment]
Looks great!

**Comment by scottschiller** [comment]
Moving the radar to render on `<canvas>` was 100% the right move, but I deferred doing this for quite a long time. 😅 Most of the battlefield is also now rendered via canvas, and the performance gains are quite significant. I also render the radar canvas at "retina" resolution, or whatever pixel density is supported by the display. On my desktop LG monitor, it's 2x.

Canvas rendering has options for retaining sharpness, e.g., on scaled images. tl;dr, I've kept the default smoothing and render the battlefield at 1x for now. I tried 2x for the battlefield sprites as well as enabling sharpening, but I found that performance took a hit on my iPhone and in the Mac OS "simulator" app. I think the difference visually is minimal, but I may make it an option for folks to try out in prefs.

Time spent in JS and rendering per-frame has dropped, so the game can now run at 60 FPS quite nicely. The higher framerate was not an initial goal, but I realized it was possible once I started the rendering refactor. Because I also implemented the game speed option recently, that helped make it easier to "re-time" the game in terms of running at 30 or 60 FPS.
